sdks/java/harness/src/test/java/org/apache/beam/fn/harness/BeamFnDataReadRunnerTest.java

```diff
@@ -122,334 +129,579 @@
-        .thenReturn(bundle2Future);
-    List<WindowedValue<String>> values = new ArrayList<>();
-    FnDataReceiver<WindowedValue<String>> consumers = values::add;
-    AtomicReference<String> bundleId = new AtomicReference<>("0");
-    BeamFnDataReadRunner<String> readRunner =
-        new BeamFnDataReadRunner<>(
-            INPUT_TRANSFORM_ID,
-            RemoteGrpcPortRead.readFromPort(PORT_SPEC, "localOutput").toPTransform(),
-            bundleId::get,
-            COMPONENTS.getCodersMap(),
-            mockBeamFnDataClient,
-            consumers);
-
-    // Process for bundle id 0
-    readRunner.registerInputLocation();
-
-    verify(mockBeamFnDataClient)
-        .receive(
-            eq(PORT_SPEC.getApiServiceDescriptor()),
-            eq(LogicalEndpoint.data(bundleId.get(), INPUT_TRANSFORM_ID)),
-            eq(CODER),
-            consumerCaptor.capture());
-
-    Future<?> future =
-        executor.submit(
-            () -> {
-              // Sleep for some small amount of time simulating the parent blocking
-              Uninterruptibles.sleepUninterruptibly(100, TimeUnit.MILLISECONDS);
-              try {
-                consumerCaptor.getValue().accept(valueInGlobalWindow("ABC"));
-                consumerCaptor.getValue().accept(valueInGlobalWindow("DEF"));
-              } catch (Exception e) {
-                bundle1Future.fail(e);
-              } finally {
-                bundle1Future.complete();
-              }
-            });
-
-    readRunner.blockTillReadFinishes();
-    future.get();
-    assertThat(values, contains(valueInGlobalWindow("ABC"), valueInGlobalWindow("DEF")));
-
-    // Process for bundle id 1
-    bundleId.set("1");
-    values.clear();
-    readRunner.registerInputLocation();
-
-    verify(mockBeamFnDataClient)
-        .receive(
-            eq(PORT_SPEC.getApiServiceDescriptor()),
-            eq(LogicalEndpoint.data(bundleId.get(), INPUT_TRANSFORM_ID)),
-            eq(CODER),
-            consumerCaptor.capture());
-
-    future =
-        executor.submit(
-            () -> {
-              // Sleep for some small amount of time simulating the parent blocking
-              Uninterruptibles.sleepUninterruptibly(100, TimeUnit.MILLISECONDS);
-              try {
-                consumerCaptor.getValue().accept(valueInGlobalWindow("GHI"));
-                consumerCaptor.getValue().accept(valueInGlobalWindow("JKL"));
-              } catch (Exception e) {
-                bundle2Future.fail(e);
-              } finally {
-                bundle2Future.complete();
-              }
-            });
-
-    readRunner.blockTillReadFinishes();
-    future.get();
-    assertThat(values, contains(valueInGlobalWindow("GHI"), valueInGlobalWindow("JKL")));
-
-    verifyNoMoreInteractions(mockBeamFnDataClient);
-  }
+  // Test different cases of chan nel split with empty allowed split points.
+  @RunWith(Parameterized.class)
+  public static class ChannelSplitTest {
```

Comment:
This test is the same to [test_simple_split](https://github.com/apache/beam/blob/master/sdks/python/apache_beam/runners/worker/bundle_processor_test.py#L61) + [test_split_with_element_progress](https://github.com/apache/beam/blob/master/sdks/python/apache_beam/runners/worker/bundle_processor_test.py#L73)